File: sorting-searching/sorting-algorithms/External-merge-sort/ExternalMergeSort.java
import java.io.*;
import java.util.*;

public class ExternalMergeSort {

    static final int CHUNK_SIZE = 4;

    static int createSortedRuns(String inputFile) throws Exception {
        Scanner scanner = new Scanner(new File(inputFile));

        List<Integer> data = new ArrayList<>();

        while (scanner.hasNextInt()) {
            data.add(scanner.nextInt());
        }

        scanner.close();

        int runIndex = 0;

        for (int i = 0; i < data.size(); i += CHUNK_SIZE) {
            List<Integer> chunk =
                data.subList(i, Math.min(i + CHUNK_SIZE, data.size()));

            List<Integer> sortedChunk = new ArrayList<>(chunk);

            Collections.sort(sortedChunk);

            PrintWriter writer =
                new PrintWriter("run" + runIndex + ".txt");

            for (int value : sortedChunk) {
                writer.print(value + " ");
            }

            writer.close();

            runIndex++;
        }

        return runIndex;
    }

    static void mergeRuns(int totalRuns, String outputFile)
        throws Exception {

        List<List<Integer>> runs = new ArrayList<>();

        for (int i = 0; i < totalRuns; i++) {
            Scanner scanner =
                new Scanner(new File("run" + i + ".txt"));

            List<Integer> run = new ArrayList<>();

            while (scanner.hasNextInt()) {
                run.add(scanner.nextInt());
            }

            scanner.close();

            runs.add(run);
        }

        PriorityQueue<Integer> minHeap =
            new PriorityQueue<>();

        List<Integer> merged = new ArrayList<>();

        for (List<Integer> run : runs) {
            minHeap.addAll(run);
        }

        while (!minHeap.isEmpty()) {
            merged.add(minHeap.poll());
        }

        PrintWriter output =
            new PrintWriter(outputFile);

        for (int value : merged) {
            output.print(value + " ");
        }

        output.close();
    }

    public static void main(String[] args)
        throws Exception {

        PrintWriter input =
            new PrintWriter("input.txt");

        input.print("8 3 7 4 9 2 6 1");

        input.close();

        int totalRuns =
            createSortedRuns("input.txt");

        mergeRuns(totalRuns, "sorted_output.txt");

        System.out.println(
            "External Merge Sort completed."
        );

        System.out.println(
            "Check sorted_output.txt"
        );
    }
}

File: sorting-searching/sorting-algorithms/External-merge-sort/external-merge-sort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <queue>
#include <algorithm>
using namespace std;

const int CHUNK_SIZE = 4;

void createSortedRuns(const string& inputFile) {
    ifstream input(inputFile);

    vector<int> chunk;
    int value;
    int runIndex = 0;

    while (input >> value) {
        chunk.push_back(value);

        if (chunk.size() == CHUNK_SIZE) {
            sort(chunk.begin(), chunk.end());

            ofstream runFile("run" + to_string(runIndex) + ".txt");

            for (int num : chunk) {
                runFile << num << " ";
            }

            runFile.close();

            chunk.clear();
            runIndex++;
        }
    }

    if (!chunk.empty()) {
        sort(chunk.begin(), chunk.end());

        ofstream runFile("run" + to_string(runIndex) + ".txt");

        for (int num : chunk) {
            runFile << num << " ";
        }

        runFile.close();
    }

    input.close();
}

void mergeRuns(int totalRuns, const string& outputFile) {
    vector<ifstream> runFiles(totalRuns);

    for (int i = 0; i < totalRuns; i++) {
        runFiles[i].open("run" + to_string(i) + ".txt");
    }

    priority_queue<
        pair<int, int>,
        vector<pair<int, int>>,
        greater<pair<int, int>>
    > minHeap;

    for (int i = 0; i < totalRuns; i++) {
        int value;

        if (runFiles[i] >> value) {
            minHeap.push({value, i});
        }
    }

    ofstream output(outputFile);

    while (!minHeap.empty()) {
        auto smallest = minHeap.top();
        minHeap.pop();

        int value = smallest.first;
        int fileIndex = smallest.second;

        output << value << " ";

        if (runFiles[fileIndex] >> value) {
            minHeap.push({value, fileIndex});
        }
    }

    output.close();

    for (auto& file : runFiles) {
        file.close();
    }
}

int main() {
    ofstream input("input.txt");

    input << "8 3 7 4 9 2 6 1";

    input.close();

    createSortedRuns("input.txt");

    mergeRuns(2, "sorted_output.txt");

    cout << "External Merge Sort completed.\n";
    cout << "Check sorted_output.txt\n";

    return 0;
}
